S3 upload artifacts

ChaiBapchya opened the pull request:
## Description ##
Upload nosetest artifacts such as nosetests_python2_cpu_quantization.xml, nosetests_python2_cputrain_.xml and nosetests_python2_cpu_unittest.xml for every job (once it is merged on the master branch)

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] ci/Jenkins_utils.groovy
- [x] Add S3 upload plugin to the jenkins-mxnet-ci.amazon.com; 
- [x] Allow read/write access to the specific S3 bucket (for Jenkins job to be able to upload to the S3 bucket
- [x] To maintain security, S3 bucket name is added as a Global property (like an environment variable of sorts) inside the Jenkins -> Configure


## Test ##

Test all the above changes if they work on the MXNet CI Dev account.
Expected result - creation of a directory of the form `<commit-id>-jenkins-<job-name>-<build-id>`
That directory should contain corresponding artifacts.

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

ChaiBapchya commented:
@marcoabreu @perdasilva @mseth10

aaronmarkham commented:
Where does the `s3Upload` function come from? This? https://github.com/jenkinsci/pipeline-aws-plugin
Also, have you put any docs together on the flow for this on the wiki or elsewhere?
Did you add the plugin to ci-dev and prod already?

ChaiBapchya commented:
Correct plugin.
Yes I tested it with dev.
Added it to prod.

Any recommendation where I should document this workflow?

aaronmarkham commented:
> Correct plugin.
> Yes I tested it with dev.
> Added it to prod.
> 
> Any recommendation where I should document this workflow?

https://cwiki.apache.org/confluence/display/MXNET/Continuous+Integration

ChaiBapchya commented:
@aar

> > Correct plugin.
> > Yes I tested it with dev.
> > Added it to prod.
> > Any recommendation where I should document this workflow?
> 
> https://cwiki.apache.org/confluence/display/MXNET/Continuous+Integration

Added the workflow here - https://cwiki.apache.org/confluence/display/MXNET/Tracking+Unit-tests

@marcoabreu I added the credentials to CI. Can you also review the cwiki. Thanks

marcoabreu commented:
Yeah we're not allowed to add AWS credentials to CI. Please remove them.

marcoabreu commented:
Okay, we clarified over Slack and Chai is in fact using instance profiles based on the jenkins-slave role paired with a policy that grants specific access to that particular bucket. That's exactly the way to go, so we're fine to move forward security-wise.

Still waiting for the exception swallow before approving.

marcoabreu commented on ci/Jenkinsfile_utils.groovy:
```diff
@@ -150,12 +149,11 @@ def collect_test_results_windows(original_file_name, new_file_name) {
-            sh "S3 Upload failed ${e}"
-          }
+        try {
+          s3Upload(file:new_file_name, bucket:env.MXNET_CI_UNITTEST_ARTIFACT_BUCKET, path:env.BRANCH_NAME+"-"+utils.get_git_commit_hash().trim()+"-"+env.BUILD_TAG+"/"+new_file_name)
```
I'm afraid that the folder starts to get too much context into its name which isn't really scalable.

I'd propose to create a context file (outside of this function since the function is called multiple times per run) where you store all the context you're interested in. The folder names should only look like "PR-12345/22" or "master/852". Everything else like commit hash, target branch, etc should then go into the context file.

ChaiBapchya replied:
Sounds good!

ChaiBapchya replied:
Resolved by discussion: JOB_NAME/PR/build/file
Because Job-level metrics are required
